tests/test_compose.py

```diff
@@ -0,0 +1,145 @@
+        dag_other = retworkx.PyDAG()
+        node_d = dag_other.add_node('d')
+        dag_other.add_child(node_d, 'e', {'a': 3})
+        res = dag.compose(dag_other, {node_c: (node_d, {'b': 1})})
+        self.assertEqual({0: 3, 1: 4}, res)
+        self.assertEqual([0, 1, 2, 3, 4], retworkx.topological_sort(dag))
+
+    def test_compose_graph_onto_digraph_error(self):
+        digraph = retworkx.PyDiGraph()
+        graph = retworkx.PyGraph()
+        with self.assertRaises(TypeError):
+            digraph.compose(graph, {})
+
+    def test_simple_graph_composition(self):
+        graph = retworkx.PyGraph()
+        node_a = graph.add_node('a')
+        node_b = graph.add_node('b')
+        graph.add_edge(node_a, node_b, {'a': 1})
+        node_c = graph.add_node('c')
+        graph.add_edge(node_b, node_c, {'a': 2})
+        graph_other = retworkx.PyGraph()
+        node_d = graph_other.add_node('d')
+        node_e = graph_other.add_node('e')
+        graph_other.add_edge(node_d, node_e, {'a': 3})
+        res = graph.compose(graph_other, {node_c: (node_d, {'b': 1})})
+        self.assertEqual({0: 3, 1: 4}, res)
+        self.assertEqual([0, 1, 2, 3, 4], graph.node_indexes())
+
+    def test_compose_digraph_onto_graph_error(self):
+        digraph = retworkx.PyDiGraph()
+        graph = retworkx.PyGraph()
+        with self.assertRaises(TypeError):
+            graph.compose(digraph, {})
+
+    def test_edge_map_func_digraph_compose(self):
+        digraph = retworkx.PyDiGraph()
+        original_input_nodes = digraph.add_nodes_from(['qr[0]', 'qr[1]'])
+        original_op_nodes = digraph.add_nodes_from(['h'])
+        output_nodes = digraph.add_nodes_from(['qr[0]', 'qr[1]'])
+        digraph.add_edge(original_input_nodes[0], original_op_nodes[0],
+                         'qr[0]')
+        digraph.add_edge(original_op_nodes[0], output_nodes[0], 'qr[0]')
+        # Setup other graph
+        other_digraph = retworkx.PyDiGraph()
+        input_nodes = other_digraph.add_nodes_from(['qr[2]', 'qr[3]'])
+        op_nodes = other_digraph.add_nodes_from(['cx'])
+        other_output_nodes = other_digraph.add_nodes_from(['qr[2]', 'qr[3]'])
+        other_digraph.add_edges_from([(input_nodes[0], op_nodes[0], 'qr[2]'),
+                                      (input_nodes[1], op_nodes[0], 'qr[3]')])
+        other_digraph.add_edges_from([
+            (op_nodes[0], other_output_nodes[0], 'qr[2]'),
+            (op_nodes[0], other_output_nodes[1], 'qr[3]')])
+
+        def map_fn(weight):
+            if weight == 'qr[2]':
+                return 'qr[0]'
+            elif weight == 'qr[3]':
+                return 'qr[1]'
+            else:
+                return weight
+
+        digraph.remove_nodes_from(output_nodes)
+        other_digraph.remove_nodes_from(input_nodes)
+        node_map = {original_op_nodes[0]: (op_nodes[0], 'qr[0]'),
+                    original_input_nodes[1]: (op_nodes[0], 'qr[1]')}
```

Comment:
Oh, that's a good point I hadn't considered that when I wrote this test, I think it might be. Especially with https://github.com/Qiskit/retworkx/pull/134 we could just remove the nodes and retain the edges. It doesn't really effect the test but when we go to use it this in terra I think that might make more sense.